pennylane/beta/queuing/circuit.py

```diff
@@ -0,0 +1,499 @@
+                if len(op.wires) > num_ops[op]:
+                    operations[wire].insert(l, None)
+
+            l += 1
+
+        observables = OrderedDict()
+        for wire in sorted(self._grid):
+            observables[wire] = list(
+                filter(
+                    lambda op: isinstance(op, qml.operation.Observable)
+                    and op.return_type is not None,
+                    self._grid[wire],
+                )
+            )
+
+            if not observables[wire]:
+                observables[wire] = [None]
+
+        return (
+            [operations[wire] for wire in operations],
+            [observables[wire] for wire in observables],
+        )
+
+    def update_node(self, old, new):
+        """Replaces the given circuit graph node with a new one.
+
+        Args:
+            old (Operator): node to replace
+            new (Operator): replacement
+
+        Raises:
+            ValueError: if the new :class:`~.Operator` does not act on the same wires as the old one
+        """
+        # NOTE Does not alter the graph edges in any way. variable_deps is not changed, _grid is not changed. Dangerous!
+        if new.wires != old.wires:
+            raise ValueError("The new Operator must act on the same wires as the old one.")
+        new.queue_idx = old.queue_idx
+        nx.relabel_nodes(self._graph, {old: new}, copy=False)  # change the graph in place
+
+    def draw(self, charset="unicode", show_variable_names=False):
+        """Draw the CircuitGraph as a circuit diagram.
+
+        Args:
+            charset (str, optional): The charset that should be used. Currently, "unicode" and "ascii" are supported.
+            show_variable_names (bool, optional): Show variable names instead of variable values.
+
+        Raises:
+            ValueError: If the given charset is not supported
+
+        Returns:
+            str: The circuit diagram representation of the ``CircuitGraph``
+        """
+        grid, obs = self.greedy_layers()
+
+        if charset not in CHARSETS:
+            raise ValueError(
+                "Charset {} is not supported. Supported charsets: {}.".format(
+                    charset, ", ".join(CHARSETS.keys())
+                )
+            )
+
+        drawer = CircuitDrawer(
+            grid,
+            obs,
+            self.wires,
+            charset=CHARSETS[charset],
+            show_variable_names=show_variable_names,
+        )
+
+        return drawer.draw()
+
+    @property
+    def diagonalizing_gates(self):
+        """Returns the gates that diagonalize the measured wires such that they
+        are in the eigenbasis of the circuit observables.
+
+        Returns:
+            List[~.Operation]: the operations that diagonalize the observables
+        """
+        rotation_gates = []
+
+        for observable in self.observables_in_order:
+            rotation_gates.extend(observable.diagonalizing_gates())
+
+        return rotation_gates

```

Comment:
Any additional value to having this method implemented twice? (once at circuit level and once at circuitgraph level)